### PR TITLE
add deepspeed zero3 support for dpo, sft

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
reward model deepspeed zero3 finetune work with https://github.com/huggingface/peft/pull/1450/files

**Comment by sywangyi** [comment]
@libinta @regisss please help review

**Review by libinta — COMMENTED** [review]
@sywangyi  do you run into any cpu/device memory issue with ddp? what's the performance comparison for ddp and deepspeed z3?

**Comment by skaulintel** [comment]
@libinta @sywangyi  I verified that I can run sft with deepspeed using 
`python ../gaudi_spawn.py --world_size 8 --use_deepspeed sft.py ...` and it works. The performance is as follows
{'train_runtime': 666.274, 'train_samples_per_second': 48.028, 'train_steps_per_second': 0.75, 'train_loss': 1.0011464239284396, 'epoch': 1.0, 'memory_allocated (GB)': 17.61, 'max_memory_allocated (GB)': 73.99, 'total_memory_available (GB)': 94.62}

compared to mpi performance.
{'train_runtime': 624.5768, 'train_samples_per_second': 51.235, 'train_steps_per_second': 0.801, 'train_loss': 0.9905898201689124, 'epoch': 1.0, 'memory_allocated (GB)': 17.61, 'max_memory_allocated (GB)': 73.99, 'total_memory_available (GB)': 94.62}

I did not pass a deepspeed config file at it doesn't seem there is an option for sft - are you planning to add this?

As for dpo.py, I tried running with 

`python ../gaudi_spawn.py --world_size 8 --use_deepspeed dpo.py ...` and passing the config file tests/configs/deepspeed_zero_3_gaudi1.json. I was not able to run getting a segmentation fault during the mapping process. Is there a different config I should be using?

**Comment by sywangyi** [comment]
SFT accept the GaudiTrainingArguments see parser = HfArgumentParser((ScriptArguments, GaudiTrainingArguments)), so you could pass deepspeed config. from your test result. zero3 is not enable since memory is apple2apple with DDP. the advantage of zero3 is reduce memory of each rank, and make large model like llama 70b lora finetune possible.
this is my SFT command for zero3
PT_HPU_MAX_COMPOUND_OP_SIZE=10 DEEPSPEED_HPU_ZERO3_SYNC_MARK_STEP_REQUIRED=1 \
python ../gaudi_spawn.py --world_size 8 --use_deepspeed sft.py \
        --model_name_or_path meta-llama/Llama-2-7b-hf \
        --deepspeed llama2_ds_zero3_config.json \
        --output_dir="./sft" \
        --max_steps=500 \
        --logging_steps=10 \
        --save_steps=100 \
        --per_device_train_batch_size=4 \
        --per_device_eval_batch_size=1 \
        --gradient_accumulation_steps=2 \
        --learning_rate=1e-4 \
        --lr_scheduler_type="cosine" \
        --warmup_steps=100 \
        --weight_decay=0.05 \
        --optim="paged_adamw_32bit" \
        --lora_target_modules "q_proj" "v_proj" \
        --bf16 \
        --remove_unused_columns=False \
        --run_name="sft_llama2" \
        --report_to=none \
        --use_habana \
        --use_lazy_mode

FOR DPO, this is my command
PT_HPU_MAX_COMPOUND_OP_SIZE=10 DEEPSPEED_HPU_ZERO3_SYNC_MARK_STEP_REQUIRED=1 \
python ../gaudi_spawn.py --world_size 8 --use_deepspeed dpo.py \
        --model_name_or_path="sft/final_merged_checkpoint" \
        --tokenizer_name_or_path=meta-llama/Llama-2-7b-hf \
        --lora_target_modules "q_proj" "v_proj" "k_proj" "out_proj" "fc_in" "fc_out" "wte" \
        --output_dir="dpo" \
        --report_to=none \
        --deepspeed llama2_ds_zero3_config.json

llama2_ds_zero3_config.json is from examples/language-modeling folder

the numproc to map the dataset is default 24, could you reduce it to see if there's segmentation fault?

File: examples/trl/dpo.py [hunk 1]
```diff
@@ -84,6 +87,7 @@ class ScriptArguments:
     seed: Optional[int] = field(
         default=0, metadata={"help": "Random seed that will be set at the beginning of training."}
     )
+    deepspeed: Optional[str] = field(default=None, metadata={"help": "the deepspeed json config file"})
 
 
 def get_stack_exchange_paired(
```
Comment (skaulintel) [inline examples/trl/dpo.py:93]: even if I reduce num_proc to 2, I keep getting a segmentation fault. However the map continues afterward. Num_proc=1 seems to run fine.

Comment (sywangyi) [reply examples/trl/dpo.py:94]: I set default to None and didnot find crash

**Comment by libinta** [comment]
@sywangyi  do you have issue running the 70b model with readme cmd for deepspeed z3?

**Comment by sywangyi** [comment]
> @sywangyi do you have issue running the 70b model with readme cmd for deepspeed z3?

with the updated cmd from README I can run 70b model for SFT and DPO

**Comment by sywangyi** [comment]
@libinta  @regisss  could we merge the PR? any issue left?

File: examples/trl/README.md [hunk 2]
```diff
@@ -49,6 +54,51 @@ There are two main steps to the DPO training process:
...
+        --remove_unused_columns=False \
+        --run_name="sft_llama2" \
+        --report_to=none \
+        --use_habana \
+        --use_lazy_mode
+    ```
+    To merge the adaptors to get the final sft merged checkpoint, we can use the `merge_peft_adapter.py` helper script that comes with TRL:
+    ```
+    python merge_peft_adapter.py --base_model_name="meta-llama/Llama-2-70b-hf" --adapter_model_name="sft" --output_name="sft/final_merged_checkpoint"
+    ```
```
Comment (libinta) [inline examples/trl/README.md:88]: Saw issue as below when running above cmd
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py", line 578, in save_pretrained
    raise ValueError(str([w.message for w in caught_warnings]))
ValueError: [UserWarning('`do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.'), UserWarning('`do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`.')]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/litang/github/438/optimum-habana/examples/trl/merge_peft_adapter.py", line 48, in <module>
    model.save_pretrained(f"{script_args.output_name}")
  File "/usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py", line 2326, in save_pretrained
    model_to_save.generation_config.save_pretrained(save_directory)
  File "/usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py", line 580, in save_pretrained
    raise ValueError(
ValueError: The generation config instance is invalid -- `.validate()` throws warnings and/or exceptions. Fix these issues to save the configuration.


Please check

Comment (sywangyi) [reply examples/trl/README.md:88]: Hi @libinta  We have tested in two fresh envs. it's ok of SFT+merged adapter. From your Traceback, seems do_sample is False. if you check https://huggingface.co/meta-llama/Llama-2-70b-hf/blob/main/generation_config.json. the do_sample is True and SFT finetune will not change the generation config. please check the llama-2-70b model in your cache, see if it's changed.

**Review by regisss — APPROVED** [review]
LGTM

File: examples/trl/README.md (hunk 2 above)
Comment (regisss) [reply examples/trl/README.md:88]: I've not got any issues running the examples. Have you managed to make it work @libinta ?

File: examples/trl/dpo.py [hunk 3]
```diff
@@ -156,18 +161,26 @@ def return_prompt_and_responses(samples) -> Dict[str, str]:
         run_name="dpo_llama2",
         use_habana=True,
         use_lazy_mode=True,
-        use_hpu_graphs_for_training=not script_args.gradient_checkpointing,
-        use_hpu_graphs_for_inference=True,
+        use_hpu_graphs_for_training=not script_args.gradient_checkpointing and (not script_args.deepspeed),
+        use_hpu_graphs_for_inference=not script_args.deepspeed,
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/trl/dpo.py:165]: does not work if graph mode is enabled for training. work if graph mode is disabled.

**Comment by regisss** [comment]
@libinta @skaulintel Have you managed to run the examples?

**Comment by skaulintel** [comment]
@sywangyi  we need the following commit c0ab0482acea745ee36260265c3247781ff0d7c7, can you merge main?

**Comment by skaulintel** [comment]
tested sft and dpo scripts with llama70b and they run. lgtm

**Comment by sywangyi** [comment]
verified in latest main with synapseAI 1.15, please help merge, Thanks @regisss

**Pull request merged**